File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
class Solution:
    def inorder(self, node, vals):
        if not node:
            return
        self.inorder(node.left, vals)
        vals.append(node.val)
        self.inorder(node.right, vals)

    def build(self, vals, l, r):
        if l > r:
            return None
        mid  = (l + r) // 2
        node = TreeNode(vals[mid])
        node.left  = self.build(vals, l, mid - 1)
        node.right = self.build(vals, mid + 1, r)
        return node

    def balanceBST(self, root):
        vals = []
        self.inorder(root, vals)
        return self.build(vals, 0, len(vals) - 1)

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
class Solution {
public:
    void inorder(TreeNode* node, vector<int>& vals) {
        if(!node) return;
        inorder(node->left, vals);
        vals.push_back(node->val);
        inorder(node->right, vals);
    }
    TreeNode* build(const vector<int>& vals, int l, int r) {
        if(l > r) return nullptr;
        int mid = (l + r) / 2;
        TreeNode* node = new TreeNode(vals[mid]);
        node->left  = build(vals, l, mid - 1);
        node->right = build(vals, mid + 1, r);
        return node;
    }
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> vals;
        inorder(root, vals);
        return build(vals, 0, (int)vals.size() - 1);
    }
};